rust/processor/src/gap_detector.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use crate::{
+    processors::{ProcessingResult, Processor, ProcessorTrait},
+    worker::PROCESSOR_SERVICE_TYPE,
+};
+use ahash::AHashMap;
+use kanal::AsyncReceiver;
+use tracing::{error, info};
+
+// Number of batches processed before gap detected
+const GAP_DETECTION_BATCH_COUNT: u64 = 50;
+// Number of seconds between each processor status update
+const UPDATE_PROCESSOR_STATUS_SECS: u64 = 1;
+
+pub struct GapDetector {
+    next_version_to_process: u64,
+    seen_versions: AHashMap<u64, ProcessingResult>,
+    last_success_batch: Option<ProcessingResult>,
+}
+
+pub struct GapDetectorResult {
+    pub next_version_to_process: u64,
+    pub num_gaps: u64,
+    pub last_success_batch: Option<ProcessingResult>,
+}
+
+impl GapDetector {
+    pub fn new(starting_version: u64) -> Self {
+        Self {
+            next_version_to_process: starting_version,
+            seen_versions: AHashMap::new(),
+            last_success_batch: None,
+        }
+    }
+
+    pub fn process_versions(
+        &mut self,
+        result: ProcessingResult,
+    ) -> anyhow::Result<GapDetectorResult> {
+        // Check for gaps
+        let gap_detected = self.has_gap(result.clone());
+        if gap_detected {
+            self.seen_versions.insert(result.start_version, result);
+            tracing::debug!("Gap detected");
+        } else {
+            // If no gap is detected, find the latest processed batch without gaps
+            self.update_prev_batch(result);
+            tracing::debug!("No gap detected");
+        }
+
+        return Ok(GapDetectorResult {
+            next_version_to_process: self.next_version_to_process,
+            num_gaps: self.seen_versions.len() as u64,
+            last_success_batch: self.last_success_batch.clone(),
+        });
+    }
+
+    fn has_gap(&self, result: ProcessingResult) -> bool {
+        self.next_version_to_process != result.start_version
+    }
+
+    fn update_prev_batch(&mut self, result: ProcessingResult) {
+        let mut new_prev_batch = result;
+        while let Some(next_version) = self.seen_versions.remove(&(new_prev_batch.end_version + 1))
+        {
+            new_prev_batch = next_version;
+        }
+        self.last_success_batch = Some(new_prev_batch.clone());
+        self.next_version_to_process = new_prev_batch.end_version + 1;
```

Comment:
if you swap the order of these, you won't need the `new_prev_batch.clone()` 

in general, try to minimize any `.clone()` if you can- it's much more expensive to copy all the data than to pass a reference to it